### PR TITLE
Add an api for customising how verifying doubles are checked

### PR DESCRIPTION
**Pull request by JonRowe**
This is a prerequisite for rspec/rspec-rails#1238

### CODE REVIEW

**Comment by JonRowe** [comment]
What does this need to finish off? I've forgotten.

**Comment by myronmarston** [comment]
> What does this need to finish off? I've forgotten.

- [ ] Consider renaming `on_declaring_verifying_double` to `verifying_double_declaration_callbacks` (or at least make the ivar and method name align). (https://github.com/rspec/rspec-mocks/pull/832/files#r21209020)
- [ ] Doc the object/module reference classes that are now publicly exposed in the callback--you said you'd do this in a separate PR, which is fine as long as we get that in 3.2. (https://github.com/rspec/rspec-mocks/pull/832/files#r21211107)
- [ ] `DirectModuleReference#target` doesn't have a spec (although one fails that implicitly uses it, but I'd like an explicit one, like we have for the other `target` implementations).  To hit that case you need to pass an anonymous class or module to `instance_double` or `class_double`.  (https://github.com/rspec/rspec-mocks/pull/832#issuecomment-65351491) As discussed in that comment, we may want to rename `DirectModuleReference` to `AnonymousModuleReference` to accurately reflect when it is used.
- [ ] `construction_spec.rb:29` has `declared` misspelled as `delcared`.
- [ ] This PR is old enough that rebasing and getting new travis builds is probably worthwhile before merging.

A few little things, nothing major.

**Comment by JonRowe** [comment]
I've done a few of the things, can you remind me where the specs for the other implementations are?

File: lib/rspec/mocks/object_reference.rb [hunk 1]
```diff
@@ -70,11 +74,15 @@ def when_loaded
     # Represents a reference to that module.
     #
     # @private
-    class DirectModuleReference < DirectObjectReference
+    class AnonymousModuleReference < DirectObjectReference
```
Comment (myronmarston) [inline lib/rspec/mocks/object_reference.rb]: The documentation comment for this class is still inaccurate ("Used when a module is passed to `class_double` or `instance_double`" -- it's actually only used when an anonymous module is passed) -- but that's not a merge blocker.  I'll plan to get to it post-merge if you don't get to it first.

**Comment by myronmarston** [comment]
> can you remind me where the specs for the other implementations are?

It was simpler just to add the spec and push myself :).

File: lib/rspec/mocks/object_reference.rb (hunk 1 above)
Comment (myronmarston) [inline lib/rspec/mocks/object_reference.rb]: I took care of this while I was at it.

**Comment by myronmarston** [comment]
Thanks for following through on this, @JonRowe.  I'm happy with this provided we don't forget to add docs for the reference classes that are now public in a separate PR before 3.2 ships.  I'll leave this for you to review and merge.

**Pull request merged**